Refactored notifications handling

mjpieters opened the pull request:
This is rather a big refactor of the notifications storage and commands, to

- eliminate bugs with pattern removing
- ensuring thread-safety (there was none)
- improved unnotify handling (patterns are matched both on equality and on regex matching)

I've included a test suite to help keep this part in working order.

mjpieters committed:
Refactor notifications

Notification fixes:

- Only list users / patterns for the current room, not all rooms
- Allow users to remove a specific pattern by treating the 'unnotify'
  argument as text as well as a regex pattern.
- Never remove the room mapping from the top level datastructure, even
  if empty. The presence of a room id is used to verify what rooms are
  supported.
- Extract the pattern from self.message.content to avoid lowercasing and
  whitespace folding.
- Normalize the pattern in the notify and unnotify commands to ensure
  consistency and avoiding showing `<code>...</code>` markup in error
  messages. This required exposing the `regex.normalize()` function for
  direct use.
- Handle any unexpected exceptions in commands as a decorator.
- Make all command actions thread-safe

General python reworking:

- Add docstrings
- Use a leading underscore for private functions / methods
- Remove unused imports
- Use textwrap.indent to handle indentation for the notifications table
- Use f-strings to interpolate values into a string
- Create empty dictionaries using {}
- Use dict.setdefault() to simplify adding users for a new regex versus
  existing.
- When traversing the nested dict structures, use dict.items() to access
  both the key and the value for further use.
- Only write out a new copy of the file when something actually changed.
- Use a dedicated logger object so it's easy to spot where a log message
  originated.
- Use a common base class for the commands to provide access to the
  notifications object and the raw argument value from the
  message.content string.

mjpieters committed:
Update readme to reflect notification changes

mjpieters committed:
Add tests for Notifications.py

mjpieters committed:
Add threading test to verify thread-safety

The test isn't foolproof, but running it with --disable-notification-locking always fails, at least on my setup.

mjpieters committed:
Mark patterns up as inline code when replyingq

mjpieters committed:
Update readme to reflect inline code markup.

tripleee commented:
Thanks, this is looking excellent! I'm not in a place where I can easily play around with this so I will probably be sitting on it for a few days. Please don't hesitate to remind me with a ping if I don't get this merged and deployed during next week.

mjpieters committed:
Build improvements: clean out test artefacts, limit pytest to Source dir

mjpieters committed:
Add a GitHub workflow to run the test suite on commits and pull requests

mjpieters committed:
Improve threading test

- Guaranteed to succeed if locking works, no more race conditions between threads
- Fails regularly when locking is neutered with --disable-notification-locking

Only a single thread has different users in different rooms use `notify ...` and `notifications` commands, and *lots* of threads running interference with a single user altering between `notify ...` and `unnotify ...`.

This then causes multiple threads to alter the dictionaries the code needs to iterate over at the same time.

mjpieters commented:
@tripleee not a problem!

I started this because I wanted to fix the constant `ValueError` exceptions I saw when trying to use `unnotify` and got a bit carried away there.

One thing I see I forgot to mention: I updated the `usage()` methods for `notify` and `unnotify` from `...` to `* ...`; they both take 1 or more arguments, not zero or more.

I've also added a simple github actions workflow to run the tests every time a commit is made to the master branch, or if a pull request against the same is made.

mjpieters committed:
Remove accidental duplicated backticks

mjpieters committed:
Clear logs earlier.

tripleee commented:
Thanks! Deployed now.

https://chat.stackexchange.com/transcript/message/54307170#54307170
